kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt

```diff
@@ -110,7 +145,24 @@ internal open class CancellableContinuationImpl<in T>(
         }
     }
 
+    /*
+     * Attempt to postpone cancellation for reusable cancellable continuation
+     */
+    private fun cancelLater(cancelled: CancelledContinuation): Boolean {
+        if (resumeMode != MODE_ATOMIC_DEFAULT) return false
+        val dispatched = (delegate as? DispatchedContinuation<*>) ?: return false
+        return dispatched.postponeCancellation(cancelled)
+    }
+
     public override fun cancel(cause: Throwable?): Boolean {
+        if (cancelLater(CancelledContinuation(this, cause, handled = state is CancelHandler))) {
```

Comment:
There are several problems here. First of all, we create an expensive `CancelledContinuation` and read `state` every time here, for no particular reason, since the `CancelledContinuation` instance will not be actually used (only `cause` is needed). Moreover, this cancel is also called from `checkCompleted` which does not need the call to `cancelLater` here at all.

The second problem is that direct call to `continuation.cancel` should not be deferred via `cancelLater`. The test in `ReusableCancellableContinuationTest.testNotCancelledOnClaimedResume` makes the wrong assumption and should be updated to reflect this fact.

Only notification from the parent via `ChildContinuation` class needs to go through `cancelLater`, so I'd suggest to remove `cancelLater` from here completely. Instead, create a separate function to call it from `ChildContinuation`:

```
    internal fun parentCancelled(cause: Throwable) {
        if (cancelLater(cause)) {
            return
        }
        cancel(cause)
    }
```

See a separate comment on reworking `cancelLater` and `postponeCancellation` so that it accepts `cause: Throwable` as parameter.

Reply:
Nice idea with storing only the throwable.

>The second problem is that direct call to continuation.cancel should not be deferred via `cancelLater`.

Fixed. I agree that now it is much simpler, though it now even more dangerous (moreover, it always was).

E.g. consider I am re-writing channels and now, for some reason, I start calling `CancellableContinuation.cancel` directly (let's omit reusability at all).

In that case, our implementation (https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt#L561) will silently stop working in very rare concurrent cases.